contrib/babelfishpg_tsql/src/linked_servers.c

```diff
@@ -0,0 +1,1024 @@
+{
+	StringInfoData buf;
+
+	initStringInfo(&buf);
+	
+	/* If error severity is greater than 10, we interpret it as a T-SQL error; otheriwse, a T-SQL info */
+	appendStringInfo(
+		&buf,
+		"TDS client library %s: Msg #: %i, Msg state: %i, Msg: %s, Server: %s, Process: %s, Line: %i, Level: %i",
+		severity > 10 ? "error" : "info",
+		error_code,
+		state,
+		error_msg ? error_msg : "",
+		svr_name ? svr_name : "",
+		proc_name ? proc_name : "",
+		line,
+		severity
+	);
+	
+	if (severity > 10)
+		ereport(ERROR,
+			(errcode(ERRCODE_FDW_UNABLE_TO_CREATE_EXECUTION),
+			errmsg("%s", buf.data)));
+	else
+	{
+		/*
+		 * We delibrately don't call the TDS report warning/info function
+		 * here because in doing so, it spews a lot of messages client side
+		 * like for database change, language change for every single connection
+		 * made to a remote server. Thus, we just log those events in the PG log
+		 * files. It would be better to atleast send the warnings client side but
+		 * currently there is no way the client libary is able to distinguish
+		 * between a warning and an informational message.
+		 *
+		 * TODO: Distinguish between WARNING and INFO
+		 */
+		ereport(INFO,
+			(errmsg("%s", buf.data)));
+	}
+
+	return 0;
+}
+
+/*
+ * Given data from TDS client library, convert it to Datum.
+ * Used for T-SQL OPENQUERY.
+ */
+Datum
+getDatumFromBytePtr(LinkedServerProcess lsproc, void *val, int datatype, int len)
+{
+	bytea *bytes;
+	
+	switch (datatype)
+	{
+		case TSQL_IMAGE:
+		case TSQL_VARBINARY:
+		case TSQL_BINARY:
+		case TSQL_BINARY_X:
+		case TSQL_VARBINARY_X:
+			bytes = palloc0(len + VARHDRSZ);
+			SET_VARSIZE(bytes, len + VARHDRSZ);
+			memcpy(VARDATA(bytes), (LS_BYTE *) val, len);
+			return PointerGetDatum(bytes);
+		case TSQL_BIT:
+		case TSQL_BITN:
+			return BoolGetDatum(*(bool *)val);
+		case TSQL_VARCHAR:
+		case TSQL_VARCHAR_X:
+		case TSQL_CHAR:
+		case TSQL_CHAR_X:
+		case TSQL_XML:
+		case TSQL_NVARCHAR_X:
+		case TSQL_NCHAR_X:
+			/*
+			 * All character data types are received from the client library in a format that can
+			 * directly be stored in a PG tuple store so they need our TDS side receiver magic.
+			 */
+			PG_RETURN_VARCHAR_P((VarChar *)cstring_to_text_with_len((char *)val, len));
+			break;
+		case TSQL_TEXT:
+		case TSQL_NTEXT:
+			/*
+			 * All character data types are received from the client library in a format that can
+			 * directly be stored in a PG tuple store, so they do not need our TDS side receiver magic.
+			 */
+			PG_RETURN_TEXT_P(cstring_to_text_with_len((char *)val, len));
+			break;
+		case TSQL_DATETIME:
+		case TSQL_DATETIMN:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr)
+			{
+				StringInfo pbuf;
+
+				pbuf = palloc(sizeof(StringInfoData));
+				/*
+				 * Rather than copying data around, we just set up a phony
+				 * StringInfo pointing to the correct portion of the TDS message
+				 * buffer. 
+				 */
+				pbuf->data = (char *)val;
```

Comment:
have we verified that it is safe to use StringInfo without '/0' in all these functions?

Reply:
We are only storing pointer to the array. The TDS receiver side logic will only use data of length specified in the buffer length/max length.